src/dotnet/ToolPackage/LocalToolsResolverCache.cs

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.Extensions.EnvironmentAbstractions;
+using Newtonsoft.Json;
+using NuGet.Frameworks;
+using NuGet.Versioning;
+
+namespace Microsoft.DotNet.ToolPackage
+{
+    internal class LocalToolsResolverCache
+    {
+        private readonly DirectoryPath _cacheVersionedDirectory;
+        private readonly IFileSystem _fileSystem;
+
+        public LocalToolsResolverCache(IFileSystem fileSystem, DirectoryPath cacheDirectory, int version)
+        {
+            _fileSystem = fileSystem ?? throw new ArgumentNullException(nameof(fileSystem));
+            _cacheVersionedDirectory = cacheDirectory.WithSubDirectories(version.ToString());
+        }
+
+        public void Save(
+            CommandSettingsListId commandSettingsListId,
+            IReadOnlyList<CommandSettings> commandSettingsList,
+            DirectoryPath nuGetGlobalPackagesFolder)
```

Comment:
And ensure nuget won't use fallback folder

Reply:
confirmed, we still use the temp project that has everything about fallback folder disabled